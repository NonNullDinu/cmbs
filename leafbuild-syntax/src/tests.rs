fn parses_without_errors(s: &str) {
    assert_eq!(crate::parser::parse(s).errors, vec![]);
}

// Those are only *parsing* tests, so as long as they make sense
// *syntactically*, they should parse.
// Even if semantically they are void of meaning.
#[test]
fn unit_to_unit_assignment() {
    parses_without_errors("() = ()\n")
}

#[test]
fn line_trailing_space() {
    parses_without_errors("() = () \n")
}

#[test]
fn if_condition() {
    parses_without_errors("if (()) {} else {}")
}

#[test]
fn if_else_if_condition() {
    parses_without_errors("if () {} else if () {} else {}")
}

#[test]
fn if_else_if_else_condition_stretched() {
    parses_without_errors(
        r#"
    if ()

    {



    }

    else if () {


    }

    else {


    }
    "#,
    )
}

#[test]
fn if_condition_in_expr() {
    parses_without_errors("() = if () {} else {} \n")
}

#[test]
fn if_condition_in_expr_stretched() {
    parses_without_errors(
        r#"
    () = if ()
    {

    }

    else
    {


    }
    "#,
    )
}

#[test]
fn declaration() {
    parses_without_errors("let x = ()\n")
}

#[test]
fn foreach_basic() {
    parses_without_errors("foreach () in () {}")
}
