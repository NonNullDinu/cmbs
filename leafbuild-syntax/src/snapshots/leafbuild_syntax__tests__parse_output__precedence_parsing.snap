---
source: leafbuild-syntax/src/tests.rs
expression: "x = 1 + 2 * 3 % - 4 ( 5 )\n"

---
- ROOT 0..26
  - Assignment 0..26
    - Expr 0..1
      - PrimaryExpr 0..1
        - "x" ID 0..1
    - " " WHITESPACE 1..2
    - "=" EQ 2..3
    - " " WHITESPACE 3..4
    - Expr 4..25
      - InfixBinOpExpr 4..25
        - PrimaryExpr 4..5
          - "1" NUMBER 4..5
        - " " WHITESPACE 5..6
        - "+" PLUS 6..7
        - " " WHITESPACE 7..8
        - InfixBinOpExpr 8..25
          - InfixBinOpExpr 8..13
            - PrimaryExpr 8..9
              - "2" NUMBER 8..9
            - " " WHITESPACE 9..10
            - "*" ASTERISK 10..11
            - " " WHITESPACE 11..12
            - PrimaryExpr 12..13
              - "3" NUMBER 12..13
          - " " WHITESPACE 13..14
          - "%" PERCENT 14..15
          - " " WHITESPACE 15..16
          - PrefixUnaryOpExpr 16..25
            - "-" MINUS 16..17
            - " " WHITESPACE 17..18
            - FuncCallExpr 18..25
              - PrimaryExpr 18..19
                - "4" NUMBER 18..19
              - " " WHITESPACE 19..20
              - FuncCallArgs 20..25
                - "(" L_PAREN 20..21
                - " " WHITESPACE 21..22
                - Expr 22..23
                  - PrimaryExpr 22..23
                    - "5" NUMBER 22..23
                - " " WHITESPACE 23..24
                - ")" R_PAREN 24..25
    - "\n" NEWLINE 25..26

